File: frontend/src/utils/invoice.ts
import { jsPDF } from 'jspdf';

interface InvoiceItem {
  name: string;
  weightSpec?: string;
  price: number;
  mrp?: number;
  qty: number;
}

interface InvoiceOrder {
  orderNumber: string;
  orderPlacedAt?: string;
  createdAt?: string;
  items: InvoiceItem[];
  itemTotal: number;
  itemTotalMrp?: number;
  deliveryFee: number;
  handlingFee: number;
  totalAmount: number;
  deliveryAddress?: string;
  paymentMethod?: string;
}

const money = (n: number) => `Rs. ${Number(n || 0).toFixed(2)}`;

/** Client-generated tax invoice / credit note as a downloadable PDF — the
 * order already contains everything needed, no server round-trip. */
export function downloadInvoice(order: InvoiceOrder) {
  const doc = new jsPDF({ unit: 'pt', format: 'a4' });
  const pageWidth = doc.internal.pageSize.getWidth();
  const marginX = 48;
  let y = 56;

  doc.setFont('helvetica', 'bold');
  doc.setFontSize(20);
  doc.setTextColor(46, 125, 50); // brand green
  doc.text('FreshCart', marginX, y);

  doc.setFontSize(11);
  doc.setTextColor(90, 90, 90);
  doc.text('Tax Invoice / Credit Note', pageWidth - marginX, y, { align: 'right' });

  y += 20;
  doc.setDrawColor(230, 230, 230);
  doc.line(marginX, y, pageWidth - marginX, y);
  y += 28;

  doc.setFont('helvetica', 'normal');
  doc.setFontSize(10);
  doc.setTextColor(60, 60, 60);
  const placedAt = order.orderPlacedAt || order.createdAt || '';
  doc.text(`Order ID: #${order.orderNumber}`, marginX, y);
  doc.text(`Date: ${placedAt}`, pageWidth - marginX, y, { align: 'right' });
  y += 16;
  if (order.paymentMethod) {
    doc.text(`Payment method: ${order.paymentMethod}`, marginX, y);
    y += 16;
  }
  if (order.deliveryAddress) {
    doc.text(`Delivery address: ${order.deliveryAddress}`, marginX, y, {
      maxWidth: pageWidth - marginX * 2,
    });
    y += 16;
  }

  y += 12;
  doc.setFont('helvetica', 'bold');
  doc.setFontSize(11);
  doc.setTextColor(20, 20, 20);
  doc.text('Item', marginX, y);
  doc.text('Qty', pageWidth - 220, y, { align: 'right' });
  doc.text('Price', pageWidth - 130, y, { align: 'right' });
  doc.text('Amount', pageWidth - marginX, y, { align: 'right' });
  y += 6;
  doc.setDrawColor(210, 210, 210);
  doc.line(marginX, y, pageWidth - marginX, y);
  y += 18;

  doc.setFont('helvetica', 'normal');
  doc.setFontSize(10);
  for (const item of order.items) {
    const lineTotal = Number(item.price) * Number(item.qty);
    doc.setTextColor(30, 30, 30);
    doc.text(item.name, marginX, y, { maxWidth: pageWidth - 300 });
    if (item.weightSpec) {
      doc.setTextColor(140, 140, 140);
      doc.setFontSize(8.5);
      doc.text(item.weightSpec, marginX, y + 11);
      doc.setFontSize(10);
    }
    doc.setTextColor(30, 30, 30);
    doc.text(String(item.qty), pageWidth - 220, y, { align: 'right' });
    doc.text(money(item.price), pageWidth - 130, y, { align: 'right' });
    doc.text(money(lineTotal), pageWidth - marginX, y, { align: 'right' });
    y += item.weightSpec ? 26 : 20;
  }

  y += 8;
  doc.setDrawColor(230, 230, 230);
  doc.line(marginX, y, pageWidth - marginX, y);
  y += 22;

  const row = (label: string, value: string, bold = false) => {
    doc.setFont('helvetica', bold ? 'bold' : 'normal');
    doc.setFontSize(bold ? 12 : 10);
    doc.setTextColor(bold ? 20 : 80, bold ? 20 : 80, bold ? 20 : 80);
    doc.text(label, pageWidth - 220, y);
    doc.text(value, pageWidth - marginX, y, { align: 'right' });
    y += bold ? 20 : 16;
  };

  row('Item Total', money(order.itemTotal));
  row('Delivery Fee', order.deliveryFee > 0 ? money(order.deliveryFee) : 'FREE');
  row('Handling Fee', order.handlingFee > 0 ? money(order.handlingFee) : 'FREE');
  y += 4;
  doc.setDrawColor(210, 210, 210);
  doc.line(pageWidth - 220, y, pageWidth - marginX, y);
  y += 20;
  row('Total Bill', money(order.totalAmount), true);

  y += 30;
  doc.setFont('helvetica', 'normal');
  doc.setFontSize(9);
  doc.setTextColor(150, 150, 150);
  doc.text('This is a system-generated invoice and does not require a signature.', marginX, y);

  doc.save(`FreshCart-Invoice-${order.orderNumber}.pdf`);
}

File: mobileapp/lib/core/utils/invoice.dart
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';
import 'package:freshcart/features/orders/data/models/order_model.dart';

String _money(num n) => 'Rs. ${n.toStringAsFixed(2)}';

/// Client-generated tax invoice / credit note as a PDF — the order already
/// has everything needed, no server round-trip. Mirrors the web app's
/// `frontend/src/utils/invoice.ts` output.
Future<void> downloadInvoice(OrderModel order) async {
  final doc = pw.Document();
  final green = PdfColor.fromInt(0xFF2E7D32);

  doc.addPage(
    pw.Page(
      pageFormat: PdfPageFormat.a4,
      margin: const pw.EdgeInsets.all(48),
      build: (context) {
        return pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.start,
          children: [
            pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                pw.Text(
                  'FreshCart',
                  style: pw.TextStyle(fontSize: 20, fontWeight: pw.FontWeight.bold, color: green),
                ),
                pw.Text('Tax Invoice / Credit Note', style: const pw.TextStyle(fontSize: 11, color: PdfColors.grey700)),
              ],
            ),
            pw.SizedBox(height: 8),
            pw.Divider(color: PdfColors.grey300),
            pw.SizedBox(height: 16),
            pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                pw.Text('Order ID: #${order.id}', style: const pw.TextStyle(fontSize: 10)),
                pw.Text(_dateLabel(order), style: const pw.TextStyle(fontSize: 10)),
              ],
            ),
            if (order.paymentMethod.isNotEmpty) ...[
              pw.SizedBox(height: 6),
              pw.Text('Payment method: ${order.paymentMethod}', style: const pw.TextStyle(fontSize: 10)),
            ],
            if (order.deliveryAddress.isNotEmpty) ...[
              pw.SizedBox(height: 6),
              pw.Text('Delivery address: ${order.deliveryAddress}', style: const pw.TextStyle(fontSize: 10)),
            ],
            pw.SizedBox(height: 18),
            pw.Table(
              columnWidths: const {
                0: pw.FlexColumnWidth(3),
                1: pw.FlexColumnWidth(1),
                2: pw.FlexColumnWidth(1.2),
                3: pw.FlexColumnWidth(1.2),
              },
              children: [
                pw.TableRow(
                  decoration: const pw.BoxDecoration(
                    border: pw.Border(bottom: pw.BorderSide(color: PdfColors.grey400)),
                  ),
                  children: [
                    _headCell('Item'),
                    _headCell('Qty', align: pw.TextAlign.right),
                    _headCell('Price', align: pw.TextAlign.right),
                    _headCell('Amount', align: pw.TextAlign.right),
                  ],
                ),
                for (final item in order.items)
                  pw.TableRow(children: [
                    pw.Padding(
                      padding: const pw.EdgeInsets.symmetric(vertical: 6),
                      child: pw.Column(
                        crossAxisAlignment: pw.CrossAxisAlignment.start,
                        children: [
                          pw.Text(item.product.name, style: const pw.TextStyle(fontSize: 10)),
                          if (item.selectedWeight.isNotEmpty)
                            pw.Text(item.selectedWeight, style: const pw.TextStyle(fontSize: 8, color: PdfColors.grey600)),
                        ],
                      ),
                    ),
                    _cell('${item.quantity}', align: pw.TextAlign.right),
                    _cell(_money(item.product.price), align: pw.TextAlign.right),
                    _cell(_money(item.totalPrice), align: pw.TextAlign.right),
                  ]),
              ],
            ),
            pw.SizedBox(height: 18),
            pw.Divider(color: PdfColors.grey300),
            pw.SizedBox(height: 8),
            _billRow('Item Total', _money(order.subtotal)),
            _billRow('Delivery Fee', order.deliveryFee > 0 ? _money(order.deliveryFee) : 'FREE'),
            _billRow('Handling Fee', order.platformFee > 0 ? _money(order.platformFee) : 'FREE'),
            if (order.discount > 0) _billRow('Discount', '- ${_money(order.discount)}'),
            pw.SizedBox(height: 4),
            pw.Divider(color: PdfColors.grey300),
            pw.SizedBox(height: 4),
            _billRow('Total Bill', _money(order.total), bold: true),
            pw.SizedBox(height: 28),
            pw.Text(
              'This is a system-generated invoice and does not require a signature.',
              style: const pw.TextStyle(fontSize: 8.5, color: PdfColors.grey500),
            ),
          ],
        );
      },
    ),
  );

  // layoutPdf (not sharePdf) is the reliable cross-platform path: on web it
  // triggers the browser's native print/"Save as PDF" dialog (a plain
  // Blob-anchor download can get silently swallowed by some browser/embed
  // contexts), and on mobile it opens the native print/share sheet.
  await Printing.layoutPdf(
    onLayout: (_) => doc.save(),
    name: 'FreshCart-Invoice-${order.id}',
  );
}

String _dateLabel(OrderModel order) {
  final d = order.date;
  return '${d.day.toString().padLeft(2, '0')}/${d.month.toString().padLeft(2, '0')}/${d.year}';
}

pw.Widget _headCell(String text, {pw.TextAlign align = pw.TextAlign.left}) => pw.Padding(
      padding: const pw.EdgeInsets.only(bottom: 6),
      child: pw.Text(text, textAlign: align, style: pw.TextStyle(fontSize: 10.5, fontWeight: pw.FontWeight.bold)),
    );

pw.Widget _cell(String text, {pw.TextAlign align = pw.TextAlign.left}) => pw.Padding(
      padding: const pw.EdgeInsets.symmetric(vertical: 6),
      child: pw.Text(text, textAlign: align, style: const pw.TextStyle(fontSize: 10)),
    );

pw.Widget _billRow(String label, String value, {bool bold = false}) => pw.Padding(
      padding: const pw.EdgeInsets.symmetric(vertical: 3),
      child: pw.Row(
        mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
        children: [
          pw.Text(label, style: pw.TextStyle(fontSize: bold ? 12 : 10, fontWeight: bold ? pw.FontWeight.bold : pw.FontWeight.normal)),
          pw.Text(value, style: pw.TextStyle(fontSize: bold ? 12 : 10, fontWeight: bold ? pw.FontWeight.bold : pw.FontWeight.normal)),
        ],
      ),
    );
